### PR TITLE
Flex dynamic storage provisioner design proposal

### PR DESCRIPTION
**Pull request by childsb**
This is a design proposal for a flex dynamic provisioner.  It depends on https://github.com/kubernetes/kubernetes/pull/17056

I will squash commits after feedback.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/20043)

### CODE REVIEW

**Comment by markturansky** [comment]
@kubernetes/rh-storage

**Comment by bgrant0607** [comment]
cc @kubernetes/goog-cluster

File: docs/design/flex-provision.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+* **Case 10**: As a Kubernetes Administrator I want to secure the node performing provisioning from the rest of the network.
+
+
+# Design
+This design covers all of the use cases above and runs on the established dynamic provisioning framework.
+
+## Creation of Provisioner
+The Storage Administrator is responsible for creating a provisioning container.  The details and contents of the provisioning container aren't important only that it creates a new Persistent Volume before exit.  This requires *kubectl* on the provisioner and a service account token passed into the container.  
+
+Storage Adminsistrators provides Kubernetes Adminstrator with a provisioning container. Kuberntes Adminstrator ensures the container is availble on appropriate nodes.  Properties passed into the provisioner could be on a secrets volume or as environment variables which can be parsed by shell scripts or binaries within the provisioning container.
```
Comment (jsafrane) [inline docs/design/flex-provision.md]: typo: Adminsistrators
typo: availble
typo: Kuberntes

File: docs/design/flex-provision.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+Dockerfile:
+
+``` Dockerfile
+FROM centos
+RUN yum update ; yum install iscsi-initiator-utils
+RUN curl www.examples.com/me/iscsi-provision.sh >> /usr/bin/iscsi-provision.sh
+ENTRYPOINT /usr/bin/iscsi-provision.sh
+```
+docker built . -t custom/iscsi-provision:latest
```
Comment (jsafrane) [inline docs/design/flex-provision.md]: typo: built -> build

**Comment by jsafrane** [comment]
Currently creation of a new PV is a two-step process
1. the provisioner creates a template of an soon-to-be-created volume.
2. the provisioner creates the volume and fills missing pieces in the Kubernetes' PV object.

Is the proposal compatible with this approach or does it just execute `/usr/bin/iscsi-provision.sh` with the PVC (and other parameters) and gets a complete PV back in one step?

File: docs/design/flex-provision.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  name: custom-isci-provisioner
+  namespace: admin
+  labels:
+    storage-class: silver
+    type: provisioner-config
+data:
+  plugin-name: kubernetes.io/flex-provisioner
+  image: custom/iscsi-provision
+  service-account-name: foo
+  service-toekn:23092390532lksf90kwelklfw
```
Comment (jsafrane) [inline docs/design/flex-provision.md]: typo: service-toekn

**Comment by childsb** [comment]
Thanks for your edits @jsafrane.

> Currently creation of a new PV is a two-step process
> 1. the provisioner creates a template of an soon-to-be-created volume.
> 2. the provisioner creates the volume and fills missing pieces in the Kubernetes' PV object.

>Is the proposal compatible with this approach or does it just execute /usr/bin/iscsi-provision.sh with the PVC (and other parameters) and gets a complete PV back in one step?

My plan is to use a small template with only the pv.name and minimum required fields, then update the template completely based on the fully created PV.  I'd also like to automate this behavior and remove the template part of the API from the provisioner. 

Ideally the "template" wouldn't be required as its clunky to impl for provisioners.

**Comment by ekozan** [comment]
link to -> https://github.com/kubernetes/kubernetes/pull/26908

**Pull request closed without merging**